src/rgw/rgw_data_sync.cc

```diff
@@ -3211,50 +4059,571 @@ int RGWBucketShardIncrementalSyncCR::operate()
+                                         SSTR( "target=" << target_bucket.value_or(rgw_bucket()) << ":source_bucket=" << source_bucket.value_or(rgw_bucket()) << ":source_zone=" << sc->source_zone)))
+{
+  if (target_bs) {
+    target_bucket = target_bs->bucket;
+  }
+  if (source_bs) {
+    source_bucket = source_bs->bucket;
+  }
+}
+
+int RGWRunBucketSourcesSyncCR::operate()
+{
+  reenter(this) {
+    yield call(new RGWGetBucketPeersCR(sync_env, target_bucket, sc->source_zone, source_bucket, &pipes, tn));
+    if (retcode < 0 && retcode != -ENOENT) {
+      tn->log(0, "ERROR: failed to read sync status for bucket");
+      return set_cr_error(retcode);
+    }
+
+    ldpp_dout(sync_env->dpp, 20) << __func__ << "(): requested source_bs=" << source_bs << " target_bs=" << target_bs << dendl;
+
+    if (pipes.empty()) {
+      ldpp_dout(sync_env->dpp, 20) << __func__ << "(): no relevant sync pipes found" << dendl;
+    }
+
+    for (siter = pipes.begin(); siter != pipes.end(); ++siter) {
+      {
+        ldpp_dout(sync_env->dpp, 20) << __func__ << "(): sync pipe=" << *siter << dendl;
+
+        source_num_shards = siter->source.get_bucket_info().num_shards;
+        target_num_shards = siter->target.get_bucket_info().num_shards;
+        if (source_bs) {
+          sync_pair.source_bs = *source_bs;
+        } else {
+          sync_pair.source_bs.bucket = siter->source.get_bucket();
+        }
+        sync_pair.dest_bs.bucket = siter->target.get_bucket();
+
+        sync_pair.handler = siter->handler;
+
+        if (sync_pair.source_bs.shard_id >= 0) {
+          num_shards = 1;
+          cur_shard = sync_pair.source_bs.shard_id;
+        } else {
+          num_shards = std::max<int>(1, source_num_shards);
+          cur_shard = std::min<int>(0, source_num_shards);
+        }
+      }
+
+      ldpp_dout(sync_env->dpp, 20) << __func__ << "(): num shards=" << num_shards << " cur_shard=" << cur_shard << dendl;
+
+      for (; num_shards > 0; --num_shards, ++cur_shard) {
+        /*
+         * use a negatvie shard_id for backward compatibility,
+         * this affects the crafted status oid
+         */
+        sync_pair.source_bs.shard_id = (source_num_shards > 0 ? cur_shard : -1);
+        if (source_num_shards == target_num_shards) {
+          sync_pair.dest_bs.shard_id = sync_pair.source_bs.shard_id;
+        } else {
+          sync_pair.dest_bs.shard_id = -1;
+        }
+
+        ldpp_dout(sync_env->dpp, 20) << __func__ << "(): sync_pair=" << sync_pair << dendl;
+
+        yield spawn(new RGWRunBucketSyncCoroutine(sc, sync_pair, tn), false);
+        while (num_spawned() > BUCKET_SYNC_SPAWN_WINDOW) {
+          set_status() << "num_spawned() > spawn_window";
+          yield wait_for_child();
+          bool again = true;
+          while (again) {
+            again = collect(&ret, nullptr);
+            if (ret < 0) {
+              tn->log(10, "a sync operation returned error");
```

Comment:
RGWRunBucketSourcesSyncCR is swallowing errors returned by RGWRunBucketSyncCoroutine, so RGWDataSyncSingleEntryCR doesn't record them in the error_repo for retries